watchtower-plugin/src/dbm.rs

```diff
@@ -216,29 +216,38 @@ impl DBM {
         &self,
         tower_id: TowerId,
         user_id: UserId,
-    ) -> Option<RegistrationReceipt> {
-        let mut stmt = self
-            .connection
-            .prepare(
-                "SELECT available_slots, subscription_start, subscription_expiry, signature
-                    FROM registration_receipts 
-                    WHERE tower_id = ?1 AND subscription_expiry = (SELECT MAX(subscription_expiry) 
-                        FROM registration_receipts 
-                        WHERE tower_id = ?1)",
-            )
-            .unwrap();
+        subscription_start: Option<u32>,
+        subscription_expiry: Option<u32>,
+    ) -> Option<Vec<RegistrationReceipt>> {
+        let mut query = "SELECT available_slots, subscription_start, subscription_expiry, signature FROM registration_receipts WHERE tower_id = ?1".to_string();
 
-        stmt.query_row([tower_id.to_vec()], |row| {
-            let slots: u32 = row.get(0).unwrap();
-            let start: u32 = row.get(1).unwrap();
-            let expiry: u32 = row.get(2).unwrap();
-            let signature: String = row.get(3).unwrap();
+        let tower_id_encoded = tower_id.to_vec();
+        let mut params: Vec<&dyn ToSql> = vec![&tower_id_encoded];
 
-            Ok(RegistrationReceipt::with_signature(
-                user_id, slots, start, expiry, signature,
-            ))
-        })
-        .ok()
+        if subscription_expiry.is_none() {
+            query.push_str(" AND subscription_expiry = (SELECT MAX(subscription_expiry) FROM registration_receipts WHERE tower_id = ?1)")
+        } else {
+            query.push_str(" AND subscription_start>=?2 AND subscription_expiry <=?3");
+            params.push(&subscription_start);
+            params.push(&subscription_expiry)
+        }
+        let mut stmt = self.connection.prepare(&query).unwrap();
+
+        Some(
+            stmt.query_map(params.as_slice(), |row| {
+                let slots: u32 = row.get(0)?;
+                let start: u32 = row.get(1)?;
+                let expiry: u32 = row.get(2)?;
+                let signature: String = row.get(3)?;
+
+                Ok(RegistrationReceipt::with_signature(
+                    user_id, slots, start, expiry, signature,
+                ))
+            })
+            .unwrap()
+            .map(|r| r.unwrap())
+            .collect(),
+        )
```

Comment:
You can simplify this:

```suggestion
        stmt.query_map(params.as_slice(), |row| {
            let slots: u32 = row.get(0)?;
            let start: u32 = row.get(1)?;
            let expiry: u32 = row.get(2)?;
            let signature: String = row.get(3)?;

            Ok(RegistrationReceipt::with_signature(
                user_id, slots, start, expiry, signature,
            ))
        })
        .unwrap()
        .map(|r| r.ok())
        .collect()
```